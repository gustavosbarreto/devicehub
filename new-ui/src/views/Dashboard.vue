<template>
  <v-row class="mt-2 ml-2" v-if="!hasStatus">
    <v-col cols="12" md="4" class="pt-0" v-for="item in items" :key="item.id">
      <div data-test="dashboard-card">
        <Card
          :id="item.id"
          :title="item.title"
          :fieldObject="item.fieldObject"
          :content="item.content"
          :icon="item.icon"
          :buttonName="item.buttonName"
          :pathName="item.pathName"
          :nameUseTest="item.nameUseTest"
          :stats="item.stats"
        />
      </div>
    </v-col>
  </v-row>
  <v-card data-test="dashboard-failed" class="mt-2 pa-4 bg-v-theme-surface" v-else>
    <p class="text-center">Something is wrong, try again !</p>
  </v-card>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref } from "vue";
import { INotificationsError } from "../interfaces/INotifications";
import Card from "../components/Card/Card.vue";
import { useStore } from "../store";

type ItemCard = {
  id: number;
  title: string;
  fieldObject: string;
  content: string;
  icon: string;
  buttonName: string;
  pathName: string;
  nameUseTest: string;
  stats: number;
};

export default defineComponent({
  name: "DashboardView",
  components: { Card },
  setup() {
    const store = useStore();
    const items = ref<ItemCard[]>([]);
    const hasStatus = ref(false);
    const itemsStats = computed(() => store.getters["stats/stats"]);

    onMounted(async () => {
      try {
        await store.dispatch("stats/get");
        items.value = [
          {
            id: 1,
            title: "Registered Devices",
            fieldObject: "registered_devices",
            content: "Registered devices into the tenancy account",
            icon: "mdi-devices",
            stats: itemsStats.value.registered_devices ?? 0,
            buttonName: "Add Device",
            pathName: "devices",
            nameUseTest: "registeredDevices-btn",
          },
          {
            id: 2,
            title: "Online Devices",
            fieldObject: "online_devices",
            content: "Devices are online and ready for connecting",
            icon: "mdi-devices",
            stats: itemsStats.value.online_devices ?? 0,
            buttonName: "View all Devices",
            pathName: "devices",
            nameUseTest: "viewOnlineDevices-btn",
          },
          {
            id: 3,
            title: "Active Sessions",
            fieldObject: "active_sessions",
            content: "Active SSH Sessions opened by users",
            icon: "mdi-devices",
            stats: itemsStats.value.active_sessions ?? 0,
            buttonName: "View all Sessions",
            pathName: "sessions",
            nameUseTest: "viewActiveSession-btn",
          },
        ];
      } catch (error: any) {
        switch (true) {
          case error.response && error.response.status === 403: {
            hasStatus.value = true;
            break;
          }
          default: {
            hasStatus.value = true;
            store.dispatch(
              "snackbar/showSnackbarErrorAction",
              INotificationsError.dashboard,
            );
            break;
          }
        }
        throw new Error(error);
      }
    });

    return {
      hasStatus,
      itemsStats,
      items,
    };
  },
});
</script>
