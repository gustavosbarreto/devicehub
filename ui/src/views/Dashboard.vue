<template>
  <fragment>
    <v-alert
      v-if="flag"
      type="warning"
      outlined
    >
      Sorry, we couldn't find the page you were looking for
    </v-alert>
    <v-row>
      <v-col
        cols="12"
        md="4"
        class="pt-0"
      >
        <v-card class="pa-2">
          <v-list-item three-line>
            <v-list-item-content>
              <div class="overline mb-4">
                Registered Devices
              </div>
              <v-list-item-title class="headline mb-1">
                {{ stats.registered_devices }}
              </v-list-item-title>
              <v-list-item-subtitle>Registered devices into the tenancy account</v-list-item-subtitle>
            </v-list-item-content>

            <v-list-item-avatar
              tile
              size="80"
            >
              <v-icon x-large>
                devices
              </v-icon>
            </v-list-item-avatar>
          </v-list-item>

          <v-card-actions>
            <DeviceAdd />
            <v-btn
              text
              data-cy="addDevice-btn"
              @click="$store.dispatch('modals/showAddDevice', true)"
            >
              Add Device
            </v-btn>
          </v-card-actions>
        </v-card>
      </v-col>

      <v-col
        cols="12"
        md="4"
        class="pt-0"
      >
        <v-card class="pa-2">
          <v-list-item three-line>
            <v-list-item-content>
              <div class="overline mb-4">
                Online Devices
              </div>
              <v-list-item-title class="headline mb-1">
                {{ stats.online_devices }}
              </v-list-item-title>
              <v-list-item-subtitle>Devices are online and ready for connecting</v-list-item-subtitle>
            </v-list-item-content>

            <v-list-item-avatar
              tile
              size="80"
            >
              <v-icon x-large>
                devices
              </v-icon>
            </v-list-item-avatar>
          </v-list-item>

          <v-card-actions>
            <v-btn
              to="/devices"
              text
              data-cy="viewDevices-btn"
            >
              View all Devices
            </v-btn>
          </v-card-actions>
        </v-card>
      </v-col>

      <v-col
        cols="12"
        md="4"
        class="pt-0"
      >
        <v-card class="pa-2">
          <v-list-item three-line>
            <v-list-item-content>
              <div class="overline mb-4">
                Active Sessions
              </div>
              <v-list-item-title class="headline mb-1">
                {{ stats.active_sessions }}
              </v-list-item-title>
              <v-list-item-subtitle>Active SSH Sessions opened by users</v-list-item-subtitle>
            </v-list-item-content>

            <v-list-item-avatar
              tile
              size="80"
            >
              <v-icon x-large>
                history
              </v-icon>
            </v-list-item-avatar>
          </v-list-item>

          <v-card-actions>
            <v-btn
              to="/sessions"
              text
              data-cy="viewSessions-btn"
            >
              View all Sessions
            </v-btn>
          </v-card-actions>
        </v-card>
      </v-col>
    </v-row>
  </fragment>
</template>

<script>
import DeviceAdd from '@/components/device/DeviceAdd.vue';

export default {
  name: 'Dashboard',

  components: {
    DeviceAdd
  },

  data(){
    return {
      flag: false
    };
  },

  computed: {
    stats() {
      return this.$store.getters['stats/stats'];
    }
  },

  created() {
    this.$store.dispatch('stats/get');
  },

  mounted() {
    this.flag = localStorage.getItem('flag');
    localStorage.removeItem('flag');
  }
};
</script>

