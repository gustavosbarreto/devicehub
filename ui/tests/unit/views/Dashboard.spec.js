import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Dashboard from '@/views/Dashboard';
import 'mock-local-storage';

describe('Dashboard', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let wrapper;
  let wrapper2;

  const numberNamespaces1 = 3;
  const numberNamespaces2 = 1;

  const namespace1 = {
    name: 'namespace1',
    owner: 'user1',
    members: [{ name: 'user3' }, { name: 'user4' }, { name: 'user5' }],
    tenant_id: 'a736a52b-5777-4f92-b0b8-e359bf484713',
  };

  const namespace2 = {
    name: 'namespace2',
    owner: 'user4',
    members: [{ name: 'user4' }, { name: 'user5' }],
    tenant_id: 'a736a52b-5777-4f92-b0b8-e359bf484712',
  };

  const statsDev = {
    registered_devices: 2,
    pending_devices: 1,
    rejected_devices: 1,
  };

  const statsNoDevices = {
    registered_devices: 0,
    pending_devices: 0,
    rejected_devices: 0,
  };

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      stats: statsDev,
      namespace: namespace1,
      numberNamespaces: numberNamespaces1,
    },
    getters: {
      'stats/stats': (state) => state.stats,
      'namespaces/getNumberNamespaces': (state) => state.numberNamespaces,
      'namespaces/get': (state) => state.namespace,
    },
    actions: {
      'stats/get': () => {
      },
      'namespaces/get': () => {
      },
      'auth/setShowWelcomeScreen': () => {
      },
      'snackbar/showSnackbarErrorLoading': () => {
      },
    },
  });

  const storeNoDevices = new Vuex.Store({
    namespaced: true,
    state: {
      stats: statsNoDevices,
      namespace: namespace2,
      numberNamespaces: numberNamespaces2,
    },
    getters: {
      'stats/stats': (state) => state.stats,
      'namespaces/getNumberNamespaces': (state) => state.numberNamespaces,
      'namespaces/get': (state) => state.namespace,
    },
    actions: {
      'stats/get': () => {
      },
      'namespaces/get': () => {
      },
      'auth/setShowWelcomeScreen': () => {
      },
      'snackbar/showSnackbarErrorLoading': () => {
      },
    },
  });

  beforeEach(() => {
    localStorage.setItem('namespacesWelcome', JSON.stringify({}));
    wrapper = shallowMount(Dashboard, {
      store,
      localVue,
      stubs: ['fragment'],
    });
  });

  it('Is a Vue instance', () => {
    expect(wrapper).toBeTruthy();
  });
  it('Renders the component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('Compare data with default value', () => {
    expect(wrapper.vm.stats.registered_devices).toBe(2);
    expect(wrapper.vm.stats.pending_devices).toBe(1);
    expect(wrapper.vm.stats.rejected_devices).toBe(1);
  });
  it('Renders the template with data', () => {
    expect(wrapper.find('[data-cy="addDevice-btn"]').exists()).toBe(true);
    expect(wrapper.find('[data-cy="viewDevices-btn"]').exists()).toBe(true);
    expect(wrapper.find('[data-cy="viewSessions-btn"]').exists()).toBe(true);
  });
  it('The welcome screen loads with the expected behavior', async () => {
    wrapper2 = shallowMount(Dashboard, { // wrapper without devices
      store: storeNoDevices,
      localVue,
      stubs: ['fragment'],
    });

    expect(wrapper2.vm.hasDevices()).toBe(false);
    await wrapper2.vm.showScreenWelcome().then(() => {
      expect(wrapper2.vm.show).toBe(true);
    });
    localStorage.setItem('namespacesWelcome', JSON.stringify({ ...{ [namespace2.tenant_id]: true } }));

    await wrapper2.vm.showScreenWelcome().then(() => {
      expect(wrapper2.vm.show).toBe(false);
    });

    expect(wrapper.vm.namespaceHasBeenShown(namespace1.tenant_id)).toBe(false);

    localStorage.setItem('namespacesWelcome', JSON.stringify({
      ...JSON.parse(localStorage.getItem('namespacesWelcome')),
      ...{ [namespace1.tenant_id]: true },
    }));

    expect(wrapper.vm.namespaceHasBeenShown(namespace1.tenant_id)).toBe(true);

    await wrapper.vm.showScreenWelcome(); // wrapper having devices
    expect(wrapper.vm.hasDevices()).toBe(true);
    expect(wrapper.vm.show).toBe(false);
    expect(Object.keys(JSON.parse(localStorage.getItem('namespacesWelcome')))).toHaveLength(2);
  });
});
