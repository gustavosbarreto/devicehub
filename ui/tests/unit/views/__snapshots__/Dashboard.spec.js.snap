// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Renders the component 1`] = `
<fragment-stub>
  <v-alert-stub color="#bd4147" outlined="true" tag="div" value="true" closelabel="$vuetify.close" dismissible="true" closeicon="$cancel" icon="" text="true">
    Sorry, we couldn't find the page you were looking for
  </v-alert-stub>
  <v-row-stub tag="div" class="mt-4">
    <v-col-stub cols="12" md="4" tag="div" class="pt-0">
      <v-card-stub loaderheight="4" tag="div" elevation="1" outlined="true">
        <v-list-item-stub activeclass="" tag="div" threeline="true">
          <v-list-item-content-stub tag="div">
            <div class="overline mb-4">
              Registered Devices
            </div>
            <v-list-item-title-stub tag="div" class="headline mb-1">
              2
            </v-list-item-title-stub>
            <v-list-item-subtitle-stub tag="div" class="grey--text">
              Registered devices into the tenancy account
            </v-list-item-subtitle-stub>
          </v-list-item-content-stub>
          <v-list-item-avatar-stub tile="true" size="80">
            <v-icon-stub xlarge="" x-large="">
              devices
            </v-icon-stub>
          </v-list-item-avatar-stub>
        </v-list-item-stub>
        <v-card-actions-stub tag="div" class="ma-2">
          <div>
            <deviceadd-stub smallbutton="true" data-test="addDevice-btn"></deviceadd-stub>
          </div>
        </v-card-actions-stub>
      </v-card-stub>
    </v-col-stub>
    <v-col-stub cols="12" md="4" tag="div" class="pt-0">
      <v-card-stub loaderheight="4" tag="div" elevation="1" outlined="true">
        <v-list-item-stub activeclass="" tag="div" threeline="true">
          <v-list-item-content-stub tag="div">
            <div class="overline mb-4">
              Online Devices
            </div>
            <v-list-item-title-stub tag="div" class="headline mb-1">
              0
            </v-list-item-title-stub>
            <v-list-item-subtitle-stub tag="div" class="grey--text">
              Devices are online and ready for connecting
            </v-list-item-subtitle-stub>
          </v-list-item-content-stub>
          <v-list-item-avatar-stub tile="true" size="80">
            <v-icon-stub xlarge="" x-large="">
              devices
            </v-icon-stub>
          </v-list-item-avatar-stub>
        </v-list-item-stub>
        <v-card-actions-stub tag="div" class="ma-2">
          <div>
            <v-btn-stub color="primary" tag="button" activeclass="" to="[object Object]" small="true" text="true" type="button" data-test="viewDevices-btn" class="v-btn--active">
              View all Devices
            </v-btn-stub>
          </div>
        </v-card-actions-stub>
      </v-card-stub>
    </v-col-stub>
    <v-col-stub cols="12" md="4" tag="div" class="pt-0">
      <v-card-stub loaderheight="4" tag="div" elevation="1" outlined="true">
        <v-list-item-stub activeclass="" tag="div" threeline="true">
          <v-list-item-content-stub tag="div">
            <div class="overline mb-4">
              Active Sessions
            </div>
            <v-list-item-title-stub tag="div" class="headline mb-1">
              0
            </v-list-item-title-stub>
            <v-list-item-subtitle-stub tag="div" class="grey--text">
              Active SSH Sessions opened by users
            </v-list-item-subtitle-stub>
          </v-list-item-content-stub>
          <v-list-item-avatar-stub tile="true" size="80">
            <v-icon-stub xlarge="" x-large="">
              devices
            </v-icon-stub>
          </v-list-item-avatar-stub>
        </v-list-item-stub>
        <v-card-actions-stub tag="div" class="ma-2">
          <div>
            <v-btn-stub color="primary" tag="button" activeclass="" to="[object Object]" small="true" text="true" type="button" data-test="viewSessions-btn" class="v-btn--active">
              View all Sessions
            </v-btn-stub>
          </div>
        </v-card-actions-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</fragment-stub>
`;
